src/core/border_router/routing_manager.cpp

```diff
@@ -913,6 +913,63 @@ void RoutingManager::SendRouterAdvertisement(const OmrPrefixArray &aNewOmrPrefix
     }
 }
 
+bool RoutingManager::IsReceivdRouterAdvertFromManager(const Ip6::Nd::RouterAdvertMessage &aRaMessage) const
+{
+    // Determines whether or not a received RA message was prepared by
+    // by `RoutingManager` itself.
+
+    bool     isFromManager = false;
+    uint16_t rioCount      = 0;
+
+    VerifyOrExit(aRaMessage.ContainsAnyOptions());
+
+    for (const Ip6::Nd::Option &option : aRaMessage)
+    {
+        if (option.GetType() == Ip6::Nd::Option::kTypePrefixInfo)
+        {
+            // PIO should match `mLocalOnLinkPrefix`.
+
+            const Ip6::Nd::PrefixInfoOption &pio = static_cast<const Ip6::Nd::PrefixInfoOption &>(option);
+            Ip6::Prefix                      prefix;
+
+            VerifyOrExit(pio.IsValid());
+            pio.GetPrefix(prefix);
+
+            VerifyOrExit(prefix == mLocalOnLinkPrefix);
+        }
+        else if (option.GetType() == Ip6::Nd::Option::kTypeRouteInfo)
+        {
+            // RIO (with non-zero lifetime) should match entries from
+            // `mAdvertisedOmrPrefixes`. We keep track of the number
+            // of matched RIOs and check after the loop ends that all
+            // entries were seen.
+
+            const Ip6::Nd::RouteInfoOption &rio = static_cast<const Ip6::Nd::RouteInfoOption &>(option);
+            Ip6::Prefix                     prefix;
+
+            VerifyOrExit(rio.IsValid());
+            rio.GetPrefix(prefix);
+
+            if (rio.GetRouteLifetime() != 0)
+            {
+                VerifyOrExit(mAdvertisedOmrPrefixes.ContainsMatching(prefix));
```

Comment:
Theoretically, `mAdvertisedOmrPrefixes` can be updated between sending the RA and receiving the loopback. Is it easier to cache the sent RA message buffer and compare the received message here byte by byte?

Reply:
In our current design, `mAdvertisedOmrPrefixes ` tracks the prefixes that were included in the last RA message we sent. We update it only after we send the RA message. We remember it till next RA message so we can know if a previously advertised prefix is later removed and can advertise it in the next RA message with zero lifetime. 

We can save the full message but since we are already saving the list of prefixes, we can reuse this. 

While I agree BRs should in general not be as memory restricted (compared to SED style devices), I think it may be better to still optimize memory use if/when we can. The situation I have in mind is RTOS style BR devices.

Reply:
I was thinking this case:
1. we sent a RA message (`RA1`) and updated the `mAdvertisedOmrPrefixes` to `prefixes1`
2. we sent a RA message (`RA2`) and updated the `mAdvertisedOmrPrefixes` to `prefixes2`
3. we received the `RA1` loopback message and compare it with `prefixes2` and we don't skip it because they are different.

But I realized my proposal won't address this issue either :) (unless we save copy of multiple RAs) This case should happen rare and may just ignore it.

Reply:
Yes. I agree we can ignore this situation. It should be rather unlikely that we sent a next RA before receiving the previous one. Even if we don't detect that a previous RA was sent by us, we keep using the RA header for a bit longer (will be fixed when we get next set of RAs).